Fix #78939 - Windows relative path with drive letter

mvorisek opened the pull request:
Fix # 78939, paths like "C:x.txt" are relative, see https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats

Please note that path like "C:" is not absolute too (tested in Windows command line by `cd C:` - cwd did not changed).

mvorisek commented:
@cmb69 Why do you think `C:` is an absolute path? According to https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats it is not absolute and if you run `cd C:` in Windows command line the path do not change to `C:` neither. `C:/` or `C:\` paths are absolute.

I will look on the tests tomorrow.

mvorisek commented:
`C:` test in C#: https://dotnetfiddle.net/y6w0Eu

mvorisek commented:
These test failed:
**- Phar - Bug #71625 - Crash in php7.dll [C:\projects\php-src\ext\phar\tests\bug71625.phpt]**
local output:
```
Warning: Phar::offsetSet(C:/phpbuild/php-src/x64/Release_TS/A:A:.phar): failed to open stream: No such file or directory in C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt on line 17

Fatal error: Uncaught PharException: unable to open new phar "C:/phpbuild/php-src/x64/Release_TS/A:A:.phar" for writing in C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt:17
Stack trace:
#0 C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt(17): Phar->offsetSet('hello_habr.txt', '<? Hello Habr!?...')
#1 {main}
  thrown in C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt on line 17
```
It seems that the bug #71625 was solved with this PR, how to proceed with this test?

**- PostgreSQL pg_fetch_*() functions [C:\projects\php-src\ext\pgsql\tests\17result.phpt]**
I do not have a PostgreSQL database, can you check?

**- recursive mkdir() tests [C:\projects\php-src\ext\standard\tests\file\mkdir-004.phpt]**
and
**- recursive mkdir() tests [C:\projects\php-src\ext\standard\tests\file\mkdir-005.phpt]**
```
chdir("/");
var_dump(getcwd());
var_dump(mkdir("./testdir/subdir", 0777, true));
var_dump(rmdir("./testdir/subdir"));
var_dump(rmdir("./testdir"));
```
outputs:
```
string(3) "C:\"

Warning: mkdir(): No such file or directory in C:/phpbuild\php-src\ext\standard\tests\file\mkdir-005.phpt on line 16
bool(false)

Warning: rmdir(./testdir/subdir): No such file or directory in C:/phpbuild\php-src\ext\standard\tests\file\mkdir-005.phpt on line 17
bool(false)

Warning: rmdir(./testdir): No such file or directory in C:/phpbuild\php-src\ext\standard\tests\file\mkdir-005.phpt on line 18
bool(false)
```
The cwd is printed correctly with the trailing backslash, but it seems that PHP internally removes the backslash and after this PR it thinks that the resulting path is no longer absolute. I have no debugging tools for C, can you please check this too?

mvorisek commented:
@cmb69 code updated, now only this test is failing:
- Phar - Bug #71625 - Crash in php7.dll [C:\projects\php-src\ext\phar\tests\bug71625.phpt]
local output:
```
Warning: Phar::offsetSet(C:/phpbuild/php-src/x64/Release_TS/A:A:.phar): failed to open stream: No such file or directory in C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt on line 17

Fatal error: Uncaught PharException: unable to open new phar "C:/phpbuild/php-src/x64/Release_TS/A:A:.phar" for writing in C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt:17
Stack trace:
#0 C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt(17): Phar->offsetSet('hello_habr.txt', '<? Hello Habr!?...')
#1 {main}
  thrown in C:\phpbuild\php-src\ext\phar\tests\bug71625.phpt on line 17
```
It seems that the bug #71625 was solved with this PR, how to update this test?

cmb69 commented:
> Why do you think C: is an absolute path?

Because I misinterpreted a quick test, and didn't read [the docs](https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats) thoroughly, which state that `C:Projects\apilibrary\apilibrary.sln` is "a relative path from the current directory of the C: drive." So apparently each drive has its own current directory. Let's check:
````.bat
@echo off

:: create test folder layout, unless already there
for %%i in (C:\test5001 C:\test5001\c D:\test5001 D:\test5001\d) do (
    if not exist %%i mkdir %%i
)

:: set cwd for each drive
cd C:\test5001
cd D:\test5001

:: verify cwd is set independently for each drive
C:
echo %CD%
D:
echo %CD%

:: check PHP's behavior
php -r "var_dump(ZEND_THREAD_SAFE)";
php -r "var_dump(file_exists('C:c'));"
php -r "var_dump(realpath('C:c'));"
````
I get
````
C:\test5001
D:\test5001
bool(true)
bool(false)
bool(false)
````
and
````
C:\test5001
D:\test5001
bool(false)
bool(true)
bool(false)
`````
So `realpath()` (and `scandir()`; I didn't check other functions) do not support drive relative paths at all; for NTS builds, the behavior is inconsistent wrt. `file_exists()`. Anyhow, PHP, and particularly TSRM, has no notion of a current working directory *per drive*, so supporting this is hardly possible, and certainly not something that we'd like to introduce in a stable version. Instead we may consider to let `file_exists()` fail for drive relative paths in NTS builds as well. And of course, we should document that.

mvorisek commented:
> So apparently each drive has its own current directory. 

According to https://devblogs.microsoft.com/oldnewthing/20101011-00/?p=12563 there should be only one current working directory and the behaviour in cmd should be considered cmd only. And the linked article (link in the previous article does not work) https://devblogs.microsoft.com/oldnewthing/20100506-00/?p=14133

So if there is only one cwd and this cwd is very well set/reset by php (otherwise standard relative paths will not work) and if relative links are properly detected (by this PR), all relative links should work if the resolution is done by the OS itself.

The different results of `file_exists` (and also of all other file functions) should be fixed by this commit to behave exactly the same between TS and NTS build (prior this PR one build used buggy `IS_ABSOLUTE_PATH`). Did you run your code with PHP with this PR patch?

Additionally, this PR fixes:
- https://github.com/php/php-src/pull/5001/files#diff-6348a66bca1dbd140e8886d403f07f40L1293 - fix no longer needed as code with the updated `IS_ABSOLUTE_PATH` is never needed as the topmost condition is never satisfied
- https://github.com/php/php-src/pull/5001/files#diff-093fc4d3c1dfbb6819809690e36128d7R1265 in recursive mkdir PHP tries to figure the first parent existing directory using specific code, it is covered by tests and I fixed it

> It seems that the bug #71625 was solved with this PR, how to update this test?
How should I fix this test? The bug is about `A:A` filename, which no longer resolves to absolute path and the nature of the bug changed - this filename is now just like any other unsuported filename (colon is not allowed in filenames on Windows). Should the test be simply updated with the new exception message or what is the standard process to fix these tests when the tested bug is no longer presented?

mvorisek commented:
@cmb69 Can you please review the changes and merge?

nikic commented on main/streams/plain_wrapper.c:
```diff
@@ -1262,7 +1262,11 @@ static int php_plain_files_mkdir(php_stream_wrapper *wrapper, const char *dir, i
+#ifdef PHP_WIN32
+			 || (offset == 3 && IS_ABSOLUTE_WIN32_LOCAL_PATH(buf, offset))
+#endif
+			) && (p = strrchr(buf, DEFAULT_SLASH))) ) {
```
I find it very hard to understand what is going on here. Can this be extracted into a helper function?

mvorisek replied:
@nikic Can you please propose the helper function? I just added an extra condition to the while loop to stop if root level of Windows path has been reached.

- the whole while condition checks if the path has more levels to strip and the path is not a root, but it is really specific for this part of the code, as it uses `oiffset` and also updates the `p` pointer.
- the `offset != 1` condition + the fix condition checks if the path is not a root, but only for normalized local file path (i.e. not UNC etc.)

cmb69 commented on Zend/zend_virtual_cwd.h:
```diff
@@ -68,10 +68,12 @@ typedef unsigned short mode_t;
 /* COPY_WHEN_ABSOLUTE is 2 under Win32 because by chance both regular absolute paths
    in the file system and UNC paths need copying of two characters */
 #define COPY_WHEN_ABSOLUTE(path) 2
```
This seems no longer true, and likely would need to be adapted.

mvorisek replied:
The only usage in PHP is on line `Zend/zend_virtual_cwd.c(1415):`.

For linux it is zero and in the code `Zend/zend_virtual_cwd.c` it is incremented then, so `2` seems like ok to me.

But the comment seems very wrong, as UNC path like `//` does not make much more sense (UNC path without host).

In this commit https://github.com/php/php-src/pull/5001/commits/e3cd4c96dfde90ea00c9f64c64b649d8da85cfa4 I tried to set it to 3.

As this definition is hard to understand, is it ok to remove it?

mvorisek replied:
@cmb69 The only usage in PHP is on line `Zend/zend_virtual_cwd.c(1415):`.

For linux it is zero and in the code `Zend/zend_virtual_cwd.c` it is incremented then, so `2` seems like ok to me.

But the comment seems very wrong, as UNC path like `//` does not make much more sense (UNC path without host).

In this commit https://github.com/php/php-src/pull/5001/commits/e3cd4c96dfde90ea00c9f64c64b649d8da85cfa4 I tried to set it to 3.

As this definition is hard to understand, is it ok to remove it?

cmb69 replied:
Not sure, but that *might* be used by (external) extensions.

mvorisek replied:
@cmb69 I just finished my experiments and the current usage of `COPY_WHEN_ABSOLUTE()` is fully ok.

mvorisek commented:
If there are no any other questions, the PR is ready to be merged.

cmb69 commented:
````php
<?php
var_dump(chdir('C:/'));
var_dump(file_exists('C:/test5001'));
var_dump(file_exists('C:test5001'));
````
prints now for NTS:
````
bool(true)
bool(true)
bool(false)
````
and for ZTS:
````
bool(true)
bool(true)
bool(true)
`````
That doesn't look like the desired output. What am I missing?

mvorisek commented:
@cmb69 Please add a test case and I will try to make the code to pass.

cmb69 commented:
Pretty strange, but I can no longer reproduce the different behavior of NTS vs. ZTS with or without this patch. Maybe that is related to the latest Windows updates I got yesterday?

Anyhow, I think it's best to start with a simple test, just to show the desired change:
````
--TEST--
Bug #78939 (Windows relative path with drive letter)
--SKIPIF--
<?php
if (PHP_OS_FAMILY !== 'Windows') die('skip this test is for Windows platforms only');
?>
--FILE--
<?php
list($drive, $topdir) = (explode('\\', __FILE__, 3));
var_dump(chdir("$drive\\"));
var_dump(file_exists("$drive\\$topdir"));
var_dump(file_exists("$drive$topdir"));
?>
--EXPECT--
bool(true)
bool(true)
bool(false)
````
This passes for me with the patch, while versions without this patch would have printed `bool(true)` three times.

And please rebase onto any of the actively supported PHP versions or master.

mvorisek commented:
@cmb69  I have added a test case.

I have also found that `file_exists` (exists functions in general) and `is_dir` (is some file type functions) is processed differently. I have identified the cause and fixed that code as well. Even if I do not like the relative Windows paths, they should now work as specified by Microsoft on both TS & NTS builds and the PR is complete now.

mvorisek commented:
@cmb69 Can we merge/close this issue?

mvorisek commented:
@cmb69 Can you review and merge to close it?

mvorisek commented:
> The patch doesn't look bad, but should have also tests to actually check whether drive relative paths work as expected for different file system functions (e.g. fopen(), file_get_contents()), but also for other functions which could be affected (cf. [3d3f11e](https://github.com/php/php-src/commit/3d3f11ede4cc7c83d64cc5edaae7c29ce9c6986f)). Also there should be tests regarding the behavior of open_basedir, but that [may not be possible](https://github.com/php/php-src/blob/php-7.4.2RC1/tests/security/open_basedir.inc). It also would be good to have which show the behavior of changing to a drive relative path on another drive, but that may also be a bit of an issue (test must be skipped if other drive is not available; how to get info of available drives?) If automated tests are not possible, this would have to be tested manually.

Standard system functions use the same underlaying path resolution. Feel free to add more tests if needed, same for the manual multidrive test if needed.

> Anyhow, neither I nor the PHP 7.4 RM @derickr would like to see this patch going into the release branches for stability reason. Therefore, please rebase onto master.

Rebased and changed the merge target. All tests passed.

mvorisek commented:
@cmb69 Can we proceed with this PR?

mvorisek commented:
@cmb69 , @nikic Is this PR welcomed?

nikic commented:
Sorry, I don't have familiarity with Windows.

weltling commented:
Thanks for the ping, @cmb69. I would recommend against implementing this feature. We might not be able to guarantee the drive relative path is safe. There are similar issues with resolving symlinks, thanks for reminding on #1243 which looks quite similar . What comes in mind that needs to be tested

- thread safe SAPI, in particular Apache with a good concurrency level
- open_basedir compatibility
- chdir() done internally in the runtime

I would at least recommend to be hesitant merging at once, until these and possibly more tests have succeeded. The discrepancy between TS/NTS and sensitivity to internal chdir() calls might be hard to solve.

Thanks.

mvorisek commented:
@weltling Appreciate your feedback. I am also more again supporting the drive relative paths as almost noone knows about this feature and when someone uses it, in a lot of cases it is by a mistake.

I have implemented this patch as there is a partial support currently (in a few places in the PHP code) and I decided to improve the support.

What do you think @weltling, @cmb69, @nikic about completely removing the support of drive relative paths and to return false from all file functions with warning? More specifically the same result as if path is outside open_basedir.

cmb69 commented:
> […] completely removing the support of drive relative paths […]

I think that would be reasonable, but probably we should first deprecate using these drive relative paths.

nikic commented:
@cmb69 I think given that a) this is very niche, b) it's already broken / support is hit and miss, and c) we're at a major version, it makes sense to remove support directly (with a nice error message). Otherwise we have to carry around this potentially dangerous functionality for the whole major version.

cmb69 commented:
@nikic well, that actually makes sense.

mvorisek commented:
> 
>  I think given that a) this is very niche, b) it's already broken / support is hit and miss, and c) we're at a major version, it makes sense to remove support directly (with a nice error message). Otherwise we have to carry around this potentially dangerous functionality for the whole major version.

> well, that actually makes sense.

I agree as well, Windows relative path with drive letter should be not supported, almost noone knows about it and thus very confusing.

Can someone update this PR to reject all Windows relative pathed with drive letter in all functions/extensions?

mvorisek commented:
closing as I have no time to remove the support as discussed in https://github.com/php/php-src/pull/5001#issuecomment-585869044
